### PR TITLE
Refactor multi-file build parsing into a single `BuildPlan` class.

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/flutter/flutter/issues/148444 (code de-duplicated).
Closes https://github.com/flutter/flutter/issues/150877 (`--build-strategy=local`).
Closes https://github.com/flutter/flutter/issues/150884 (`--build-strategy=remote`).

Replaces duplicate code across ~3 commands (query, test, build) with a class called `BuildPlan`, which encapsulates (a) concurrency, (b) build configuration, (c) RBE, (d) LTO, and (e) build strategy. I also moved all of the validation of the build plan into `build_plan_test`, which gives us better coverage at a lower cost (less things to think about in that suite of tests).

I know the diff looks scary, but 1K of the 1.4K is tests.

/cc @gaaclarke @flar

### CODE REVIEW

File: tools/engine_tool/lib/src/build_plan.dart [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+import 'package:args/args.dart';
+import 'package:collection/collection.dart';
+import 'package:engine_build_configs/engine_build_configs.dart';
+import 'package:meta/meta.dart';
+
+import 'build_utils.dart';
+import 'environment.dart';
+import 'logger.dart';
+
+const _flagConfig = 'config';
```
Comment (johnmccutchan) [inline tools/engine_tool/lib/src/build_plan.dart:14]: Why are we pulling flags out of flags.dart?

**Review by johnmccutchan — APPROVED** [review]
This is a nice cleanup

File: tools/engine_tool/lib/src/build_plan.dart (hunk 1 above)
Comment (matanlurey) [reply tools/engine_tool/lib/src/build_plan.dart:14]: They are no longer visible or relevant to the commands (they are implementation details).

Comment (matanlurey) [reply tools/engine_tool/lib/src/build_plan.dart:14]: That is, the reason we had them in a common place no longer applies.

**Pull request merged**